### ISSUE TITLE
Laptop muxless nvidia card don't work with 2 displays on PRIME Offload

### ISSUE THREAD

**Issue by AdsonCicilioti** [NixOS/nixpkgs#189378]
### Describe the bug
I still don't have the knowledge to understand how everything works on the "Nix side of the force", but I believe the way the drivers and settings/options to cover device types could improve. Because I don't understand how a Laptop with a muxless GPU can't work in Prime Offload mode, where the integrated graphics work in primary mode, but the Nvidia dGPU is available for Xorg and applications that require it.

On Arch Linux just install the driver, configure some kernel modules (`nvidia` `nvidia_uvm` `nvidia_modeset` `nvidia_drm`) and enable the kernel parameter `options nvidia_drm modeset=1`.
Ready! Laptop will use iGpu all the time. A second HDMI monitor managed directly by the nvidia dGPU will also work just fine.

On Nix I can't reproduce this behavior. I have no advanced knowledge, but here is the part of my `/etc/configuration.nix` file that handles the video issues:

### Steps To Reproduce
```
boot = {
   kernelModules = [ "nvidia" "nvidia_uvm" "nvidia_drm" "nvidia_modeset" ];
   extraModulePackages = [ config.boot.kernelPackages.nvidia_x11 ];
   extraModprobeConfig = ''
      options nvidia NVreg_UsePageAttributeTable=1
      options nvidia NVreg_RegistryDwords="OverrideMaxPerf=0x1"
      options nvidia NVreg_PreserveVideoMemoryAllocations=1 NVreg_TemporaryFilePath=/var/tmp
   '';
   kernelPackages = pkgs.linuxPackages_zen;
};

services.xserver = {
    enable = true;
    videoDrivers = [ "nvidia" ];

    displayManager = {
      gdm = {
        enable = true;
        wayland = false;
      };
   };
   desktopManager.gnome.enable = true;
};

hardware.nvidia.modesetting.enable = true;
hardware.nvidia.powerManagement.enable = true;
hardware.nvidia.prime = {
    ofload.enable = true;
    intelBusId = "PCI:0:2:0";
    nvidiaBusId = "PCI:1:0:0";
};

hardware = {
    opengl = {
      enable = true;
      driSupport = true;
      driSupport32Bit = true;
    };
};
```

### Expected behavior
iGpu as primary graphics and dGpu nvidia making second HDMI/DP monitor work and other applications and games that need it.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
- system: `"x86_64-linux"`
 - host os: `Linux 5.18.14-zen1, NixOS, 22.05 (Quokka), 22.05.2676.b9fd420fa53`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.1`
 - channels(balllaz): `"home-manager-22.05.tar.gz"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by AdsonCicilioti** [comment]
I've got it using an TWM (Tiling Window Manager) and an X init.

Putting a `xrandr` command before call the TWM session:

```
xrandr --setprovideroutputsource 1 0

# AND

xrandr --output HDMI-1-0 --primary --mode 1920x1080 --pos 1920x0 --rotate normal --output eDP-1 --mode 1920x1080 --pos 0x0 --rotate normal
```

**Comment by moritzschaefer** [comment]
@AdsonCicilioti also took me a loong while to figure out the necessity for the xrandr call. I use `xrandr --setprovideroutputsource 1 0`, but your call works as well.

Unfortunately, I experience X crashes when I change the display configuration. When I first switch on a screen (`xrandr --output DP-1-2.3 --auto --left-of eDP-1`), everything works fine, but when I modify the configuration, including when I switch off the screen (`xrandr --output DP-1-2.3 --off`), my X Server crashes. I have so far not been able to find any debug logs that help pinpoiting the source of the crash.

This crashing behaviour is replicable in both KDE and EXWM.

Did you (@AdsonCicilioti) experience crashes at any point experimenting with xrandr and your muxless nvidia card? I'm running on a Thinkpad X1 Extreme G4

**Comment by AdsonCicilioti** [comment]
> Did you (@AdsonCicilioti) experience crashes at any point experimenting with xrandr and your muxless nvidia card? I'm running on a Thinkpad X1 Extreme G4

Exactly! If stay change video settings with `xrandr` in some times, the X crashes.

I'm suspicious, and slightly suspicious, of the way the nvidia driver is implemented in nixos. I investigated the nvidia derivation of nixpkgs, and it bothers me a lot that there is some Xorg config file in `/etc/X11/`. On Arch Linux for example, just install the nvidia driver, indicate the modules `nvidia`, `nvidia-drm`, `nvidia-modeset` and `nvidia-uvm` in a kernel module manager like `mkinitcpio`, for example . One important thing to improve fluidity is to add the `nvidia_drm.modeset=1` option to the kernel parameters. Ready! Any muxless laptop with Nvidia is able to work with external displays in prime offload (intel on primary) mode.

One thing I don't understand either, but I suspect it could also be causing a problem, is the `udev` rules that the `nvidia` branch implements.

Anyway. It must be a combination of things. We know that the proprietary closed `nvidia` driver doesn't help much, but it has improved and many old workrounds are no longer needed. I think the implementation should be simple, without Xorg confs, as I said above. A refactoring that makes the basic installation leaner and some settings optional would be interesting.

**Comment by pongo1231** [comment]
Hitting the issue here too on my laptop with an Intel + NVIDIA setup. In addition I seem to be unable to pass any arguments through the X config to the driver which prevents me from enabling coolbits to enable overclocking. I'm trying to enable coolbits by using this config:

```
Section "Device"
        Identifier      "iGPU"
        Driver          "intel"
        BusID          "PCI:0:2:0"
EndSection

Section "Screen"
        Identifier      "iGPU"
        Device         "iGPU"
EndSection

Section "Device"
        Identifier      "dGPU"
        Driver          "nvidia"
        BusID          "PCI:1:0:0"
        Option         "Coolbits" "24"
EndSection

Section "Screen"
        Identifier      "dGPU"
        Device         "dGPU"
        Option         "AllowEmptyInitialConfiguration"
EndSection

```

This works fine on Arch if I put it in `/etc/X11/xorg.conf.d` but doesn't seem to have any effect in NixOS even if I attempt to override xorg.conf by using `services.xserver.config = lib.mkForce ...`.

Also it seems like `nvidia-settings` is missing some options on this setup too. This is what it looks like under NixOS:

![image](https://user-images.githubusercontent.com/4201956/197294173-b9e73c0e-307d-4d63-9012-31cdeb92ead9.png)

And what it's supposed to look like:

![image](https://user-images.githubusercontent.com/4201956/197294208-59db0b30-cf6f-45fa-a93e-a1c346331623.png)

**Comment by usmcamp0811** [comment]
Still kind of new to NixOS and I have a Thinkpad P1 Gen4 experiencing these same problems; just curious if anyone has gotten this working?

**Comment by moritzschaefer** [comment]
I just tried again (X1 Extreme Gen4) and observed the same problem as others: 
- first call to xrandr works successfully and external screens show as expected
- second call crashes
